### PR TITLE
Store Beacon keys in Wallet

### PR DESCRIPTION
**Pull request by tomasbrod**
Beacon keys are stored in wallet (4 th time writing this sentence).
No keys, not even public key need to be in config.
Keys in config are ignored.
Public key is always obtained from network (appcache).
Beacon advertise stores keys in wallet too. <- **needs testing**
To use your existing beacon, the keys need to be imported.

Any key can be imported using ~~a new `importbeaconsecretkey`~~ _existing_ `importprivkey` command. ~~This command by default uses keys from config, however the 236 character key can be specified explicitly.~~ This command accepts compressed key, but I added support for 236 character hex keys.

### CODE REVIEW

**Comment by tomasbrod** [comment]
I tried to add support for automatic import.

File: src/beacon.cpp [hunk 1]
```diff
@@ -19,69 +20,55 @@ namespace
     }
 }
 
-bool GenerateBeaconKeys(const std::string &cpid, std::string &sOutPubKey, std::string &sOutPrivKey)
+bool GenerateBeaconKeys(const std::string &cpid, CKey outPrivPubKey)
```
Comment (denravonska) [inline src/beacon.cpp]: `outPrivPubKey` needs to be a reference.

File: src/beacon.h [hunk 2]
```diff
@@ -5,6 +5,7 @@
 #pragma once
 
 #include "fwd.h"
+#include "key.h"
```
Comment (denravonska) [inline src/beacon.h]: You can put `CKey` in `fwd.h` and move the include to the source file.

**Comment by sedited** [comment]
Can you create a unit test in test/key_tests.cpp (or create a new file) to check SignMessage and subsequent VerifyCPIDSignature ?

**Comment by tomasbrod** [comment]
I should add Locks there, or at lease `AssertLockHeld`.

File: src/beacon.cpp [hunk 3]
```diff
@@ -19,69 +20,55 @@ namespace
...
-    return true;
-}
+    CKeyID keyID = newKey.GetID();
 
-void StoreBeaconKeys(
-        const std::string &cpid,
-        const std::string &pubKey,
-        const std::string &privKey)
-{
-    WriteKey("publickey" + cpid + GetNetSuffix(), pubKey);
```
Comment (denravonska) [inline src/beacon.cpp]: This means that `WriteKey` is no longer in use and can be removed. Sweet!

**Comment by tomasbrod** [comment]
Tests for the most important stuff (Advertise, Import, sign and verify) are still missing.

File: src/main.cpp [hunk 4]
```diff
@@ -4768,6 +4768,10 @@ void GridcoinServices()
                 msMiningErrors6 = _("Unable To Send Beacon! Unlock Wallet!");
             }
         }
+        else
+        {
+            ImportBeaconKeysFromConfig();
```
Comment (denravonska) [inline src/main.cpp]: I'll move this to the wallet init. Is that ok?

File: src/rpcblockchain.cpp [hunk 5]
```diff
@@ -813,6 +848,9 @@ bool AdvertiseBeacon(std::string &sOutPrivKey, std::string &sOutPubKey, std::str
             return false;
         }
 
+        // here try to import keys from config
+        ImportBeaconKeysFromConfig();
```
Comment (denravonska) [inline src/rpcblockchain.cpp]: If we move the import to the init we can remove this call and have advertise just advertise.

**Comment by denravonska** [comment]
I just realized my changes won't work if the wallet is locked. I'll have a look.

File: src/beacon.cpp [hunk 6]
```diff
@@ -225,3 +201,38 @@ bool VerifyBeaconContractTx(const CTransaction& tx)
...
+    std::string sBeaconPublicKey = GetBeaconPublicKey(GlobalCPUMiningCPID.cpid,false);
+    std::string sCPID(msPrimaryCPID);
+    std::string strSecret= GetArgument("privatekey" + sCPID + (fTestNet ? "testnet" : ""), "");
+    if(strSecret.empty())
+        return false;
+    auto vecsecret = ParseHex(strSecret);
+
+    CKey key;
+    if(!key.SetPrivKey(CPrivKey(vecsecret.begin(),vecsecret.end())))
+        return error("ImportBeaconKeysFromConfig: Invalid private key");
```
Comment (denravonska) [inline src/beacon.cpp]: Hmm, the caller won't be able to distinguish between success (true) or an error (not null -> true).

File: src/beacon.cpp [hunk 7]
```diff
@@ -3,96 +3,69 @@
...
+    AssertLockHeld(pwalletMain->cs_wallet);
+    CPubKey oldKey(ParseHex(GetBeaconPublicKey(cpid, false)));
+    if(oldKey.IsValid())
+    {
+        if (pwalletMain->GetKey(oldKey.GetID(), outPrivPubKey))
+        {
+            return outPrivPubKey.IsValid();
+        }
+    }
+    return false;
```
Comment (denravonska) [inline src/beacon.cpp:68]: We need to fall back on config stored keys here for when the import hasn't taken place yet due to a locked wallet.

File: src/beacon.cpp (hunk 6 above)
Comment (denravonska) [reply src/beacon.cpp]: Don't review when tired. `error` returns false.

File: src/init.cpp [hunk 8]
```diff
@@ -795,6 +796,9 @@ bool AppInit2(ThreadHandlerPtr threads)
         else
             strErrors << _("Error loading wallet.dat") << "\n";
     }
+    
+    // here try to import keys from config
+    ImportBeaconKeysFromConfig();
```
Comment (tomasbrod) [inline src/init.cpp]: This will not work, because the wallet is locked.

**Comment by tomasbrod** [comment]
I do not really like your additions @denravonska 
EDIT: most of them are fine, just some may not work.

**Comment by denravonska** [comment]
@tomasbrod Which ones and why?

**Comment by tomasbrod** [comment]
Just the change where import is called. And I changed it to what I think is better b65bcaf. Can't be imported on start-up, because the wallet is locked then. I added comments to code.

* [during beacon timer](https://github.com/gridcoin-community/Gridcoin-Research/pull/1088/commits/b65bcafb88d2d64fc5d8c084f2c10b1d37d98615#diff-7ec3c68a81efff79b6ca22ac1f1eabbaR4773)
* [in advertisebeacon rpc](https://github.com/gridcoin-community/Gridcoin-Research/pull/1088/commits/b65bcafb88d2d64fc5d8c084f2c10b1d37d98615#diff-3627f02371334f49dce3bd06627e5967R1105)

**Comment by jamescowens** [comment]
What did we do to fix the automatic import of the beacon key(s) in the wallet? I don't think ImportBeaconKeysFromConfig is scoped to the right conditions. Most people are going to need this executed when they initially fire the wallet up after upgrading to a version with this in there, not at the "advertisebeacon" point.

I had to import the private key manually using importprivkey to get it going.

**Comment by denravonska** [comment]
@tomasbrod Very good idea!

**Pull request merged**